### PR TITLE
Improve user-control

### PR DESCRIPTION
**Pull request by jwmueller**
* Fixes LearningWithNoisyLabels on dataset with rare labels. Should address:
https://github.com/cleanlab/cleanlab/issues/51
https://github.com/cleanlab/cleanlab/issues/91

* Increases flexibility of LearningWithNoisyLabels by optionally allowing any kwarg of `find_label_issues()` to be specified there and passing kwargs into `clf.fit()`. Partially addresses (although full solution is nontrivial):
https://github.com/cleanlab/cleanlab/issues/110

* Adds warnings for edge-cases like when find_label_issues purposefully does not flag some issues in rare classes.

* Improves user-control over many previously hardcoded thresholds and non-settable kwargs throughout.

* Improves documentation language/formatting, with more cautionary statements about out-of-sample pred_probs.

### CODE REVIEW

File: cleanlab/count.py [hunk 1]
```diff
@@ -656,9 +667,24 @@ def estimate_confident_joint_and_cv_pred_proba(
         X_train_cv, X_holdout_cv = X[cv_train_idx], X[cv_holdout_idx]
         s_train_cv, s_holdout_cv = labels[cv_train_idx], labels[cv_holdout_idx]
 
-        # Fit the clf classifier to the training set and
-        # predict on the holdout set and update pred_probs.
-        clf_copy.fit(X_train_cv, s_train_cv)
+        # Ensure no missing classes in training set.
+        train_cv_classes = set(s_train_cv)
+        all_classes = set(range(K))  # TODO: more efficient if committed to labels 0,...,K, else use this: set(labels)
+        if len(train_cv_classes) != len(all_classes):
```
Comment (anishathalye) [inline cleanlab/count.py:673]: When is this possible? Unless there is only a single data point of a particular class in the entire dataset, shouldn't every fold of the stratified k-fold cross val have at least one data point of each class in the training split? And if there's only a single data point of a particular class in the entire dataset, handling it in this way doesn't seem great; that single point will be in both the train and test set. Might cause downstream issues, too; the threshold for the class is going to be equal to the predicted probability for that data point, and then by this logic (https://github.com/cleanlab/cleanlab/blob/master/cleanlab/count.py#L388), there will be no data that ends up in that entry on the diagonal of the confident joint, and ...

(the consequences / fix might be easier to discuss in-person)

**Review by anishathalye — COMMENTED** [review]
Left some inline comments. Other comments:

- Could you also add regression tests for the issues that this PR fixes, like datasets without enough example per class, etc.?
- Note that the Black format check is failing.

**Comment by anishathalye** [comment]
It would also be helpful to collect which currently open issues are resolved by this PR, extract a test case from each of the issues, and take note of the issue numbers so we can close them once this PR is merged.

File: cleanlab/filter.py [hunk 2]
```diff
@@ -242,18 +242,19 @@ def find_label_issues(
...
     You can return ONLY the indices of the label issues in your dataset, by setting
-    return_indices_ranked_by = {`self_confidence`, `normalized_margin`}.
+    `return_indices_ranked_by` to a string value specifying which label quality score to order the label issues
+    (and corresponding indices) by.
 
-    number of indices returned is specified by frac_noise. When
-    frac_noise = 1.0, all "confident" estimated noise indices are returned.
-    * If you encounter the error 'pred_probs is not defined', try setting n_jobs = 1.
+    The number of indices returned is controlled by `frac_noise`, reduce its value to identify fewer label issues.
+    Tip: If you encounter the error 'pred_probs is not defined', try setting `n_jobs = 1`.
```
[the remark this answers is not in the document]
Comment (jwmueller) [reply cleanlab/filter.py:257]: I'm not sure what this error is about, any insights @cgnorthcutt?

File: cleanlab/classification.py [hunk 3]
```diff
@@ -394,17 +378,16 @@ class label.
...
-            confident_joint=self.confident_joint,
-            filter_by=self.filter_by,
-            n_jobs=self.n_jobs,
+            **find_label_issues_kwargs,         
         )
 
         x_mask = ~self.label_issues_mask
         x_cleaned = X[x_mask]
         labels_cleaned = labels[x_mask]
+        clf_kwargs.update(clf_final_kwargs)
```
[the remark this answers is not in the document]
Comment (jwmueller) [reply cleanlab/classification.py:387]: Good point, added this and deepcopied the other dicts to avoid mutating them.

File: cleanlab/classification.py [hunk 4]
```diff
@@ -251,10 +211,16 @@ def fit(
         thresholds=None,
         noise_matrix=None,
         inverse_noise_matrix=None,
+        find_label_issues_kwargs = {"filter_by": "prune_by_noise_rate",
+                                    "min_examples_per_class": 10},
```
[the remark this answers is not in the document]
Comment (jwmueller) [reply cleanlab/classification.py:215]: Could you elaborate on this?  

The user can just solely specify `find_label_issues_kwargs={'min_examples_per_class': 5}` and the other args will be left at their defaults. These are just optional kwargs, so any kwarg not in this dict will be left at its default value in `find_label_issues()`.  I've added clarifying sentence to docstring:

"Arguments not specified here will remain at their default values."

[the remark this answers is not in the document]
Comment (jwmueller) [reply cleanlab/classification.py:215]: Also FYI: in the default setting of `find_label_issues_kwargs` I've set:
`"filter_by": "prune_by_noise_rate"` which is unnecessary. I did this to encourage users to play with it as it may be important hyperparameter, and it provides more of an example of how to use `find_label_issues_kwargs`. We can also remove this is you strongly prefer?

File: cleanlab/count.py [hunk 5]
```diff
@@ -734,6 +761,9 @@ def estimate_py_noise_matrices_and_cv_pred_proba(
         Set the default state of the random number generator used to split
         the cross-validated folds. If None, uses np.random current random state.
 
+    clf_kwargs : dict, default = {}
+          Optional keyword arguments to pass into `clf` fit() method.
```
Comment (JohnsonKuan) [inline cleanlab/count.py:765]: Can change indentation to 2 spaces

File: cleanlab/count.py (hunk 1 above)
Comment (jwmueller) [reply cleanlab/count.py:673]: Yes there is a duplicated copy of this one point in both train and test set (the test set copy is the only that matters though). The duplicate train set point is necessary to ensure model training does not fail (model's training set needs to have all classes for many model-types), and it should have minimal impact on the model.

I've updated the code to overwrite the model's prediction for the test set point. So now we use a hard-coded dummy-prediction for the duplicated test point in each crossval fold. For now, dummy prediction = one-hot vector with 1 in the entry for the given label. I've verified this at least allows LearningWithNoisyLabels to avoid crashing and does not introduce noticeable performance difference AFAICT.

File: cleanlab/classification.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (anishathalye) [reply cleanlab/classification.py:215]: This patch currently has the weird/unintuitive behavior that if you don't pass `find_label_issues_kwargs`, it will use the default argument (214--215 above), which includes `min_examples_per_class=10`, but if you pass e.g. `find_label_issues_kwargs={'filter_by': 'both'}`, when it calls `find_label_issues`, it won't pass `min_examples_per_class` and hence use _its_ default value for that argument, which is different (it's set to `1` in this PR: https://github.com/cleanlab/cleanlab/pull/149/commits/e1d9ae4fcf95e802f4a8e112845929ee785705eb#diff-9b65a96b3894947ee2eee0fd92220c97695cc9cb72a27fb256cd6fe4d31e975bR245).

[the remark this answers is not in the document]
Comment (jwmueller) [reply cleanlab/classification.py:215]: I see your point, what's your preferred way to resolve this?  They all seem to have other tradeoffs.

[the remark this answers is not in the document]
Comment (anishathalye) [reply cleanlab/classification.py:215]: Couple variations on exactly how to write it. One example:

```python
DEFAULT_FIND_LABEL_ISSUES_KWARGS = {"filter_by": "prune_by_noise_rate", "min_examples_per_class": 10}}
# document that the above is effectively the default, but you can override it component-wise
def fit(...
    find_label_issues_kwargs = {}
):
    find_label_issues_kwargs = {**DEFAULT_FIND_LABEL_ISSUES_KWARGS, **find_label_issues_kwargs}
```

[the remark this answers is not in the document]
Comment (jwmueller) [reply cleanlab/classification.py:215]: Ya that's what I was thinking, don't think users will be confused by (or even notice) this.

File: cleanlab/count.py (hunk 5 above)
Comment (jwmueller) [reply cleanlab/count.py]: Thanks, done. We'll need to standardize the docstring indentation anyway across the methods/files, but that'll be saved for a separate big PR.

File: cleanlab/filter.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jwmueller) [reply cleanlab/filter.py:291]: After looking into it, still not able to reproduce error. I don't think this is a message we are ever printing to user, it's just an error that may come up somehow due to the concurrency.

**Comment by jwmueller** [comment]
@anishathalye Latest commit should address your concerns. Also added test of a dataset without enough examples per class.

File: tests/test_classification.py [hunk 6]
```diff
@@ -105,6 +118,52 @@ def test_rp(sparse):
...
+            clf=LogisticRegression(multi_class="auto", solver="lbfgs", random_state=SEED)
+        )
+        rp.fit(
+            data["X_train"],
+            data["labels"],
+            find_label_issues_kwargs={"return_indices_ranked_by": "self_confidence"},
+        )
+    except Exception as e:
+        assert "not supported" in str(e)
+
```
Comment (anishathalye) [inline tests/test_classification.py:144]: Same as above, add else to ensure that exception is thrown

File: tests/test_classification.py [hunk 7]
```diff
@@ -105,6 +118,52 @@ def test_rp(sparse):
...
+    find_label_issues_kwargs = {"confident_joint": confident_joint, "min_examples_per_class": 2}
+    rp.fit(
+        data["X_train"],
+        data["labels"],
+        find_label_issues_kwargs=find_label_issues_kwargs,
+        clf_kwargs={},
+        clf_final_kwargs={},
+    )
+    score = rp.score(data["X_test"], data["true_labels_test"])
+    assert True
```
Comment (anishathalye) [inline tests/test_classification.py:164]: Unnecessary `assert True`

Comment (jwmueller) [reply tests/test_classification.py:164]: removed

File: tests/test_classification.py [hunk 8]
```diff
@@ -105,6 +118,52 @@ def test_rp(sparse):
...
+    data = make_rare_label(DATA)
+    test_rp(data)
+
+
+def test_invalid_inputs():
+    data = make_data(sparse=False, sizes=[1, 1, 1])
+    try:
+        test_rp(data)
+    except Exception as e:
+        assert "Need more data" in str(e)
```
[the remark this answers is not in the document]
Comment (jwmueller) [reply tests/test_classification.py:132]: good catch, added

File: tests/test_classification.py (hunk 6 above)
Comment (jwmueller) [reply tests/test_classification.py:144]: good catch, added

File: cleanlab/count.py [hunk 9]
```diff
@@ -86,7 +93,7 @@ def num_label_issues(
 
     Returns
     -------
-        An integer estimating the number of label issues."""
+        An integer estimate the number of label issues."""
```
[the remark this answers is not in the document]
Comment (jwmueller) [reply cleanlab/count.py:96]: added!

File: cleanlab/filter.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jwmueller) [reply cleanlab/filter.py:291]: @cgnorthcutt any ideas?

File: cleanlab/classification.py [hunk 10]
```diff
@@ -313,24 +303,35 @@ class label.
             t = np.round(np.trace(inverse_noise_matrix), 2)
             raise ValueError("Trace(inverse_noise_matrix) is {}. Must exceed 1.".format(t))
 
+        self._process_label_issues_kwargs(find_label_issues_kwargs)
+        clf_final_kwargs = {**clf_kwargs, **clf_final_kwargs}
+        self.clf_kwargs = deepcopy(clf_kwargs)
```
[the remark this answers is not in the document]
Comment (jwmueller) [reply cleanlab/classification.py:308]: was being extra safe :), removed the copy operation

File: cleanlab/filter.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (anishathalye) [reply cleanlab/filter.py:291]: This seems orthogonal to this PR, so let's create a new issue for it or track it internally. Doesn't have to block this PR.

**Review by anishathalye — APPROVED** [review]
lgtm

File: cleanlab/filter.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jwmueller) [reply cleanlab/filter.py:291]: Added issue to track it:  https://github.com/cleanlab/cleanlab/issues/150
